File: 2-medium/00695-max-area-of-island/solution.py
class Solution:
    def backtrack(self, grid, i, j):
        if grid[i][j] == 0:
            return

        self.count += 1
        grid[i][j] = 0

        if i > 0:
            self.backtrack(grid, i - 1, j)

        if i < len(grid) - 1:
            self.backtrack(grid, i + 1, j)

        if j > 0:
            self.backtrack(grid, i, j - 1)

        if j < len(grid[i]) - 1:
            self.backtrack(grid, i, j + 1)

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        max_area = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 1:
                    self.count = 0
                    self.backtrack(grid, i, j)
                    max_area = max(max_area, self.count)

        return max_area

File: 2-medium/00695-max-area-of-island/solution.rs
impl Solution {
    fn backtrack(grid: &mut Vec<Vec<i32>>, i: usize, j: usize, count: &mut i32) {
        if grid[i][j] == 0 {
            return;
        }

        *count += 1;
        grid[i][j] = 0;

        if i > 0 {
            Self::backtrack(grid, i - 1, j, count);
        }
        if i < grid.len() - 1 {
            Self::backtrack(grid, i + 1, j, count);
        }

        if j > 0 {
            Self::backtrack(grid, i, j - 1, count);
        }
        if j < grid[0].len() - 1 {
            Self::backtrack(grid, i, j + 1, count);
        }
    }

    pub fn max_area_of_island(mut grid: Vec<Vec<i32>>) -> i32 {
        let mut max = 0;
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if grid[i][j] == 1 {
                    let mut count = 0;
                    Self::backtrack(&mut grid, i, j, &mut count);
                    max = max.max(count);
                }
            }
        }

        max
    }
}
